atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/TransactionCacheValueStoreImpl.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.api.cache;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Sets;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.CellReference;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.keyvalue.api.cache.TransactionCacheValueStoreImpl.LocalCacheEntry.Status;
+import com.palantir.common.streams.KeyedStream;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import javax.annotation.concurrent.NotThreadSafe;
+import org.immutables.value.Value;
+
+@NotThreadSafe
+final class TransactionCacheValueStoreImpl implements TransactionCacheValueStore {
+    private final ValueCacheSnapshot snapshot;
+    private final Map<CellReference, LocalCacheEntry> localUpdates;
+
+    TransactionCacheValueStoreImpl(ValueCacheSnapshot snapshot) {
+        this.snapshot = snapshot;
+        this.localUpdates = new HashMap<>();
+    }
+
+    @Override
+    public boolean isWatched(TableReference tableReference) {
+        return snapshot.isWatched(tableReference);
+    }
+
+    @Override
+    public void cacheRemoteWrite(TableReference tableReference, Cell cell, CacheValue value) {
+        CellReference cellReference = CellReference.of(tableReference, cell);
+        if (snapshot.isWatched(tableReference) && snapshot.isUnlocked(cellReference)) {
+            localUpdates.put(cellReference, LocalCacheEntry.write(value));
+        }
+    }
+
+    @Override
+    public void cacheRemoteReads(TableReference tableReference, Map<Cell, byte[]> remoteReadValues) {
+        if (snapshot.isWatched(tableReference)) {
+            KeyedStream.stream(remoteReadValues)
+                    .mapKeys(cell -> CellReference.of(tableReference, cell))
+                    .filterKeys(snapshot::isUnlocked)
```

Comment:
No, this means that a lock has been taken out on this cell. If I understand correctly, that's supposed to mean that we cannot read this value from the cache because it is currently being changed (because some other transaction holds the lock here)?

Reply:
Yeah I was asking more around why this being locked means we cannot cache. But now that I think about it some more, since we have succeeded on a transactional read, the transaction that held the lock is already done by this point, but we do not know if we have seen the old or the new version, so we do not want to cache